File: base64_url.cs
// base64_url.cs — C# версия

using System;
using System.IO;
using System.Text;
using System.Windows.Forms;

class Base64UrlEncoder {
    static void Main(string[] args) {
        string input = "";
        string filePath = "";
        bool decode = false;
        string output = "";
        bool noCopy = false;

        for (int i = 0; i < args.Length; i++) {
            if (args[i] == "--input") {
                input = args[++i];
            } else if (args[i] == "--file") {
                filePath = args[++i];
            } else if (args[i] == "--decode") {
                decode = true;
            } else if (args[i] == "--output") {
                output = args[++i];
            } else if (args[i] == "--no-copy") {
                noCopy = true;
            } else if (!args[i].StartsWith("-")) {
                if (string.IsNullOrEmpty(input) && string.IsNullOrEmpty(filePath)) {
                    input = args[i];
                }
            }
        }

        Console.WriteLine("\u001B[36m🔐 Base64 URL-safe Encoder (C#)\u001B[0m");

        string result = "";

        if (!string.IsNullOrEmpty(filePath)) {
            Console.WriteLine($"📂 Обработка файла: {filePath}");
            if (decode) {
                string content = File.ReadAllText(filePath, Encoding.UTF8);
                result = Encoding.UTF8.GetString(Convert.FromBase64String(content.Trim()));
                // URL-safe: стандартный Convert.FromBase64String принимает и URL-safe, так как '-' и '_' не поддерживаются, нужно заменить
                // В C# нет встроенного URL-safe, поэтому заменяем вручную
                string fixedContent = content.Replace('-', '+').Replace('_', '/');
                // Добавляем padding
                int padding = 4 - (fixedContent.Length % 4);
                if (padding != 4) fixedContent += new string('=', padding);
                result = Encoding.UTF8.GetString(Convert.FromBase64String(fixedContent));
            } else {
                byte[] data = File.ReadAllBytes(filePath);
                string base64 = Convert.ToBase64String(data);
                // URL-safe: заменяем + на -, / на _, убираем =
                result = base64.Replace('+', '-').Replace('/', '_').TrimEnd('=');
            }
        } else if (!string.IsNullOrEmpty(input)) {
            Console.WriteLine($"📝 Входные данные: {input}");
            if (decode) {
                string fixedInput = input.Replace('-', '+').Replace('_', '/');
                int padding = 4 - (fixedInput.Length % 4);
                if (padding != 4) fixedInput += new string('=', padding);
                result = Encoding.UTF8.GetString(Convert.FromBase64String(fixedInput));
            } else {
                string base64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(input));
                result = base64.Replace('+', '-').Replace('/', '_').TrimEnd('=');
            }
        } else {
            Console.WriteLine("📝 Чтение из STDIN (Ctrl+D для окончания)");
            string line;
            var sb = new StringBuilder();
            while ((line = Console.ReadLine()) != null) {
                sb.AppendLine(line);
            }
            string data = sb.ToString();
            if (string.IsNullOrEmpty(data)) {
                Console.WriteLine("\u001B[33m⚠️ Пустой ввод.\u001B[0m");
                Environment.Exit(1);
            }
            if (decode) {
                string fixedData = data.Trim().Replace('-', '+').Replace('_', '/');
                int padding = 4 - (fixedData.Length % 4);
                if (padding != 4) fixedData += new string('=', padding);
                result = Encoding.UTF8.GetString(Convert.FromBase64String(fixedData));
            } else {
                string base64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(data));
                result = base64.Replace('+', '-').Replace('/', '_').TrimEnd('=');
            }
        }

        Console.WriteLine($"\u001B[32mРезультат:\u001B[0m");
        Console.WriteLine(result);
        Console.WriteLine();

        if (!noCopy) {
            if (CopyToClipboard(result)) {
                Console.WriteLine("\u001B[32m✅ Результат скопирован в буфер обмена!\u001B[0m");
            } else {
                Console.WriteLine("\u001B[33m⚠️ Не удалось скопировать в буфер обмена.\u001B[0m");
            }
        }

        if (!string.IsNullOrEmpty(output)) {
            File.WriteAllText(output, result, Encoding.UTF8);
            Console.WriteLine($"\u001B[32m💾 Сохранено в {output}\u001B[0m");
        }
    }

    private static bool CopyToClipboard(string text) {
        try {
            Clipboard.SetText(text);
            return true;
        } catch {
            return false;
        }
    }
}

File: base64_url.java
// base64_url.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.util.Base64;
import java.awt.*;
import java.awt.datatransfer.*;

public class base64_url {
    public static void main(String[] args) throws Exception {
        String input = "";
        String filePath = "";
        boolean decode = false;
        String output = "";
        boolean noCopy = false;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--input")) {
                input = args[++i];
            } else if (args[i].equals("--file")) {
                filePath = args[++i];
            } else if (args[i].equals("--decode")) {
                decode = true;
            } else if (args[i].equals("--output")) {
                output = args[++i];
            } else if (args[i].equals("--no-copy")) {
                noCopy = true;
            } else if (!args[i].startsWith("-")) {
                if (input.isEmpty() && filePath.isEmpty()) {
                    input = args[i];
                }
            }
        }

        System.out.println("\u001B[36m🔐 Base64 URL-safe Encoder (Java)\u001B[0m");

        String result = "";

        if (!filePath.isEmpty()) {
            System.out.println("📂 Обработка файла: " + filePath);
            if (decode) {
                String content = new String(Files.readAllBytes(Paths.get(filePath)), "UTF-8");
                result = new String(Base64.getUrlDecoder().decode(content.trim()));
            } else {
                byte[] data = Files.readAllBytes(Paths.get(filePath));
                result = Base64.getUrlEncoder().withoutPadding().encodeToString(data);
            }
        } else if (!input.isEmpty()) {
            System.out.println("📝 Входные данные: " + input);
            if (decode) {
                result = new String(Base64.getUrlDecoder().decode(input));
            } else {
                result = Base64.getUrlEncoder().withoutPadding().encodeToString(input.getBytes("UTF-8"));
            }
        } else {
            System.out.println("📝 Чтение из STDIN (Ctrl+D для окончания)");
            StringBuilder sb = new StringBuilder();
            BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line).append("\n");
            }
            String data = sb.toString();
            if (data.isEmpty()) {
                System.out.println("\u001B[33m⚠️ Пустой ввод.\u001B[0m");
                System.exit(1);
            }
            if (decode) {
                result = new String(Base64.getUrlDecoder().decode(data.trim()));
            } else {
                result = Base64.getUrlEncoder().withoutPadding().encodeToString(data.getBytes("UTF-8"));
            }
        }

        System.out.println("\u001B[32mРезультат:\u001B[0m");
        System.out.println(result);
        System.out.println();

        if (!noCopy) {
            if (copyToClipboard(result)) {
                System.out.println("\u001B[32m✅ Результат скопирован в буфер обмена!\u001B[0m");
            } else {
                System.out.println("\u001B[33m⚠️ Не удалось скопировать в буфер обмена.\u001B[0m");
            }
        }

        if (!output.isEmpty()) {
            Files.write(Paths.get(output), result.getBytes("UTF-8"));
            System.out.println("\u001B[32m💾 Сохранено в " + output + "\u001B[0m");
        }
    }

    private static boolean copyToClipboard(String text) {
        try {
            StringSelection selection = new StringSelection(text);
            Toolkit.getDefaultToolkit().getSystemClipboard().setContents(selection, null);
            return true;
        } catch (Exception e) {
            return false;
        }
    }
}
